Fixes #29819 - Support byos images on azurerm

apuntamb opened the pull request:
BYOS images on azure are a different type of marketplace images. They just need an added parameter from `PurchasePlan` class of the sdk. This PR provisions the host using rhel byos image if the image terms have been accepted. Refer the [MSFT doc](https://docs.microsoft.com/en-us/azure/virtual-machines/workloads/redhat/byos#requirements-and-conditions-to-access-the-red-hat-gold-images) to accept image terms.
Currently, I have to test if with this change the already supported images still function correctly or not hence, adding WIP label.  Will remove the label once verified that it works smoothly for every single image category.

vijay8451 commented:
@ShimShtein sure will give a try once have this enable over my Azure cloud account .

apuntamb commented:
Tested VM creation as per Vijay's directions as he couldn't get the byos images activated for his az account.

vijay8451 commented:
Test Results:
Works:
- Able to create/edit/delete BYOS image using `marketplace://` prefix
- Able to create/edit/delete marketplace image using `marketplace://` prefix
- Able to create/edit/delete shared gallery image using `gallery://` prefix
- Able to create/edit/delete custom image using `custom://` prefix

Found a minor issue:
- Able to create duplicate images using same URN (by just changing the case)

i.e. you could create two images using below URNs and the only diff in them is `R/r` in `RedHat`
```
marketplace://RedHat:rhel-byos:rhel-lvm81:8.1.2019123006
marketplace://redHat:rhel-byos:rhel-lvm81:8.1.2019123006
```

@ShimShtein @apuntamb thanks ACK from my end .

apuntamb commented:
> Test Results:
> Works:
> 
>     * Able to create/edit/delete BYOS image using `marketplace://` prefix
> 
>     * Able to create/edit/delete marketplace image using `marketplace://` prefix
> 
>     * Able to create/edit/delete shared gallery image using `gallery://` prefix
> 
>     * Able to create/edit/delete custom image using `custom://` prefix
> 
> 
> Found a minor issue:
> 
>     * Able to create duplicate images using same URN (by just changing the case)
> 
> 
> i.e. you could create two images using below URNs and the only diff in them is `R/r` in `RedHat`
> 
> ```
> marketplace://RedHat:rhel-byos:rhel-lvm81:8.1.2019123006
> marketplace://redHat:rhel-byos:rhel-lvm81:8.1.2019123006
> ```

Good catch @vijay8451! I have fixed this too in my latest commit.
> 
> @ShimShtein @apuntamb thanks ACK from my end .

@ShimShtein I have added case sensitive validation in the engine file for this.
Also, mentioned byos-images in the Image create form.

ShimShtein commented on lib/foreman_azure_rm/engine.rb:
```diff
@@ -42,7 +42,7 @@ class Engine < ::Rails::Engine
 
       # Add format validation for azure images
-      ::Image.validates :uuid, format: { with: /\A((marketplace|custom|gallery):\/\/)[^:]+(:[^:]+:[^:]+:[^:]+)?\z/,
+      ::Image.validates :uuid, uniqueness: { case_sensitive: false }, format: { with: /\A((marketplace|custom|gallery):\/\/)[^:]+(:[^:]+:[^:]+:[^:]+)?\z/,
```
This means you cannot add the same image ID to different Azure compute resources.
You should use `scope` to limit the uniqueness. Example: https://github.com/theforeman/foreman/blob/develop/app/models/image.rb#L13

apuntamb replied:
Thanks @ShimShtein for the catch! Made changes.

ShimShtein commented on app/models/concerns/foreman_azure_rm/vm_extensions/managed_vm.rb:
```diff
@@ -59,6 +59,24 @@ def define_data_disks(vm_name, data_disks)
 
+      def marketplace_image_plan(image)
+        image_type, image_id = image.split('://')
+        if image_type == 'marketplace' && image_id.include?('byos')
```
Nitpick: 
```suggestion
        return nil unless image_type == 'marketplace' && image_id.include?('byos')
```

apuntamb replied:
done.

ShimShtein commented:
Merged, thanks @apuntamb!
